src/Catalog/Dnx/DnxCatalogCollector.cs

```diff
@@ -0,0 +1,121 @@
+        }
+
+        public Uri ContentBaseAddress { get; set; }
+
+        protected override async Task<bool> OnProcessBatch(CollectorHttpClient client, IEnumerable<JToken> items, JToken context, DateTime commitTimeStamp, CancellationToken cancellationToken)
+        {
+            foreach (JToken item in items)
+            {
+                string id = item["nuget:id"].ToString().ToLowerInvariant();
+                string version = item["nuget:version"].ToString().ToLowerInvariant();
+                string type = item["@type"].ToString().Replace("nuget:", Schema.Prefixes.NuGet);
+
+                if (type == Schema.DataTypes.PackageDetails.ToString())
+                {
+                    // Add/update package
+                    string nuspec = await LoadNuspec(client, id, version, cancellationToken);
+                    if (nuspec != null)
+                    {
+                        Uri requestUri = new Uri(ContentBaseAddress, string.Format("{0}.{1}.nupkg", id, version));
+                        using (Stream stream = await client.GetStreamAsync(requestUri))
+                        {
+                            await _dnxMaker.AddPackage(stream, nuspec, id, version, cancellationToken);
+                        }
+
+                        Trace.TraceInformation("commit: {0}/{1}", id, version);
+                    }
+                    else
+                    {
+                        Trace.TraceWarning("no nuspec available for {0}/{1} skipping", id, version);
+                    }
+                }
+                else if (type == Schema.DataTypes.PackageDelete.ToString())
+                {
+                    await _dnxMaker.DeletePackage(id, version, cancellationToken);
+
+                    Trace.TraceInformation("commit delete: {0}/{1}", id, version);
+                }
+            }
+
+            return true;
+        }
+
+        private async Task<string> LoadNuspec(HttpClient client, string id, string version, CancellationToken cancellationToken)
+        {
+            Uri requestUri = new Uri(ContentBaseAddress, string.Format("{0}.{1}.nupkg", id, version));
+            HttpResponseMessage httpResponseMessage = await client.GetAsync(requestUri, cancellationToken);
+            if (httpResponseMessage.IsSuccessStatusCode)
+            {
+                using (Stream stream = await httpResponseMessage.Content.ReadAsStreamAsync())
+                {
+                    string nuspec = GetNuspec(stream, id);
+                    return nuspec;
+                }
+            }
+            else if (httpResponseMessage.StatusCode == System.Net.HttpStatusCode.NotFound)
+            {
+                Trace.TraceInformation("package not found...");
+            }
+            else
+            {
+                httpResponseMessage.EnsureSuccessStatusCode();
+            }
+            return null;
+        }
+
+        private static string GetNuspec(Stream stream, string id)
```

Comment:
That would be great to have the same logic in the server, use *PackageArchiveReader*
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs#L83

As you can see in the above method the client does not allow packages that have multiple nuspecs, but here the server just takes the first one and goes with it. I would expect the gallery to be even more strict on this. Bad packages can still come from other sources, but nuget.org shouldn't allow users to upload packages that cannot work.

I think users would appreciate the extra checks, I don't think anyone really wants to upload a package that gets indexed in this way since if the server handles it wrong they then cannot delete it.